blog: the best tools for product engineers (#4898)
PostHog/posthog.com: contents/blog/product-engineer-tools.md

---
date: 2022-12-23

> title: The 6 essential tools of product engineers

Reviewer: ```suggestion
title: The essential tools used by product engineers
```

Decided a number implies a listicle format. This work for you?

---
date: 2022-12-14

> title: The best tools for product engineers

Reviewer: Given there isn't much search volume specifically around tools for product engineers atm, I think we could use a more shareable headline that still works for SEO. Something like...

- x tools product engineers swear by
- The X tools product engineers can't live without
- The essential tools for product-minded engineers

Thoughts?

Author: The 6 essential tools of product engineers?

rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
author: ["ian-vanagas"]
featuredImage: ../images/blog/posthog-engineering-blog.png
featuredImageType: full
categories: ["Startups", "Guides"]
---

> Building tools for product engineers is a big trend right now. Every week brings a flashy launch of a new tool intersecting product and engineering. This is because building a great product users value is a proven way to succeed as a business. Helping product engineers make this happen is valuable.

Reviewer: ```suggestion
Building tools for [product engineers](/blog/what-is-a-product-engineer/) is a big trend right now. Every week brings a flashy launch of a new tool intersecting product and engineering. This is because building a great product users value is a proven way to succeed as a business. Helping product engineers make this happen is valuable.
```

Reviewer: For SEO reasons, always link on the earlier reference in an article, even if you have a link in the footer.

Tools for product engineers differ from both the tools for software engineers and product managers. The engineering tools focus on speed and automation rather than the traditional customization and stability. The product tools are streamlined for technical users, rather than being an all-encompassing “business tool.”

> This post will go over the tools built and used to help product engineers accomplish their work of building great products for users, from the beginning stages of talking to users all the way to maintaining ownership over issues and bugs in production.

Reviewer: Can you get this intro down to two paras? Best to keep intros on these types of articles as tight and actionable as possible. If one para is possible, even better.

> ## Talking to users

Reviewer: ```suggestion
## Tools for talking to users
```

Product engineers are more focused on users than traditional software engineers are. This means they must talk to them. The tools for doing this aren’t groundbreaking, but they are useful for accomplishing product engineers’ goal of building product users value.

To talk with users, product engineers use tools like Zoom, Google Meet, Slack (especially shared channels), [Superhuman](https://superhuman.com/), [Cron](https://cron.com/) (calendar), Calendly, and [SavvyCal](https://savvycal.com/). They also take advantage of automation by using tools like [customer.io](https://customer.io/), aha.io, [n8n.io](https://n8n.io/), and Zapier.

Unlike product managers, product engineers don’t use survey tools. The data received from surveys is too indirect to be useful. It might be valuable for product and marketing teams to get a general idea of users, but product engineers focus on direct feedback on what they are building.

> ## Get data on how users are using the product

Reviewer: ```suggestion
## Tools for understanding user behavior
```

For example, a product engineer sets an OKR to increase feature usage by paid users. They need data to both measure the usage of the feature, and areas surrounding the feature. With this, they can analyze conversion to using the feature, feature retention, correlated behavior to feature usage, and more.

[PostHog](/product#top-features) has a suite of product analytics tools that are perfect for this. Product engineers (like [Y Combinator’s](/customers/ycombinator)) use PostHog to capture, organize, analyze, and visualize product usage data. PostHog also has session recordings for going into the details of an individual’s experience with the product.

On top of PostHog, product engineers often use customer data platform or ETL tools like [Segment](/docs/integrate/third-party/segment), [Hightouch](/blog/hightouch-posthog-reverse-etl-integration), [Airbyte](/docs/apps/airbyte-export), Integrate.io, and Fivetran to move and combine data between sources and storage locations. 

> ## Using knowledge to prioritize properly and stay on track

Reviewer: ```suggestion
## Tools for organizing knowledge and priorities
```

Once product engineers talk with users and analyze usage data, they should be starting to form ideas on what is impactful to build. All this research and analysis is wasted if it does not lead to shipping better products.

Product engineers use note-taking and task management tools to help them with this. Many of these tools can weigh teams down and introduce unnecessary work that gets in the way of shipping products that help users (**cough** Jira). The tools for product engineers ideally don’t do this.

An example specifically built for product engineers is [Linear](https://linear.app/). Linear focuses on helping teams build great products. It does this by having engineer-focused design and functionality, while still being a product tool. In practice, this means being keyboard first, streamlined for technical use cases, and fast loading and usage speeds.

> ![Linear](../images/blog/product-engineer-tools/linear.png)

Reviewer: Feels a bit weird this is the only tool we have a screenshot for. Could we maybe have one screenshot per section with a caption for each?

Also, any chance we could get a good screenshot of the product rather than the website?

Author: Added more images throughout. I chose that Linear image because it explained why Linear was good for product engineers, and a screenshot of the product might not. Can change it is you disagree.

Like the “talking to users” tools, many of these tools won’t be shocking to you. Other note-taking and task management tools product engineers use include GitHub, Google Docs, Trello, Notion, Asana, and Coda. 

> ## Going fast with CI/CD and testing

Reviewer: ```suggestion
## Tools for going fast with CI/CD and testing
```

Tools used for this include GitHub, GitLab, CircleCI, Jenkins, Docker, [Depot](https://depot.dev/), [codefresh](https://codefresh.io/), and more. All of these tools help engineering teams ship code to end users faster, perfect for what product engineers are trying to do.

Because they ship fast, product engineers also must make sure they aren’t breaking things. They likely don’t have the time or resources for a dedicated QA team so they rely heavily on testing in its place. Test frameworks like [Jest](https://jestjs.io/), [Cypress](https://www.cypress.io/), and [Playwright](https://playwright.dev/) are all critical for product engineers to continue moving fast.

Even with all this preventative work, breaking changes can still happen. Feature flags help prevent this. Luckily, if you’re using PostHog for analytics, you can also use it for feature flags to help roll out and roll back changes quickly.

> ## Automating away infrastructure work

Reviewer: ```suggestion
## Tools for automating away infrastructure work
```

Many infrastructure-as-a-service companies exist, including the big ones like Google Cloud, AWS, Cloudflare, Heroku, and Azure. Product engineers might be inclined to use services like [Netlify](https://www.netlify.com/), [Render](https://render.com/), [fly.io](http://fly.io), or [Railway](https://railway.app/) which focus on frontend and web products as well as the speed of deployment rather than depth of customization.

A trend that appeals to product engineers is bringing the writing of code and deployment of code closer together. The best example of this is [Vercel](https://vercel.com/), which also develops the [Next.js](https://nextjs.org/) framework popular with product engineers. Vercel helps deploy highly performant frontend code fast and is tightly integrated with Next.js to maximize benefits from its design.

Other examples of these code and infrastructure combo include [Supabase](https://supabase.com/), [PlanetScale](https://planetscale.com/), and [Hasura](/customers/hasura). They abstract away infrastructure work while providing speed.  Software engineers and large teams might find their limited customization and options constraining.

> ## Ownership over bugs and incidents

Reviewer: ```suggestion
## Tools for owning bugs and incidents
```
